### PR TITLE
[WIP] bulk support

### PR DESCRIPTION
**Pull request by v-zhuravlev**

### CODE REVIEW

**Comment by v-zhuravlev** [comment]
to implement #21

**Comment by v-zhuravlev** [comment]
👏

**Comment by i-ky** [comment]
Unit tests for parser:
```
#include "src/datatype.h"

#define GOOD(syntax)														\
void good_##syntax(const char *datatype, int need)										\
{																\
	int	got;														\
	char	*error = NULL;													\
																\
	if (-1 == (got = parse_datatype(syntax, datatype, &error)))								\
	{															\
		printf("FAILED: \"%s\" is expected to be valid expression, but yields error: \"%s\".\n", datatype, error);	\
		free(error);													\
	}															\
	else if (got != need)													\
	{															\
		printf("FAILED: \"%s\" is expected to yield %d, not %d.\n", datatype, need, got);				\
	}															\
	else															\
		printf("OK:     \"%s\" yields %d, as expected.\n", datatype, got);						\
}

#define BAD(syntax)														\
void bad_##syntax(const char *datatype, const char *message)									\
{																\
	int	got;														\
	char	*error = NULL;													\
																\
	if (-1 == (got = parse_datatype(syntax, datatype, &error)))								\
	{															\
		if (NULL != message && 0 != strcmp(error, message))								\
			printf("FAILED: \"%s\" is expected to fail with \"%s\", not \"%s\".\n", datatype, message, error);	\
		else														\
			printf("OK:     \"%s\" is invalid, as expected: \"%s\".\n", datatype, error);				\
																\
		free(error);													\
	}															\
	else															\
		printf("FAILED: \"%s\" is expected to be invalid, but yields %d.\n", datatype, got);				\
}

GOOD(bit_syntax)
GOOD(register_syntax)
BAD(bit_syntax)
BAD(register_syntax)

/* make them shorter! */
#define good_bits	good_bit_syntax
#define bad_bits	bad_bit_syntax
#define good_regs	good_register_syntax
#define bad_regs	bad_register_syntax

int main(void)
{
	good_bits("b", 1);
	good_bits("bit", 1);
	good_bits("2*b", 2);
	good_bits("b*2", 2);
	good_bits(" 2*b ", 2);
	good_bits("2*b + 3*skip + 4*bit", 9);

	bad_bits("B", NULL);
	bad_bits("b*", NULL);
	bad_bits("*b", NULL);
	bad_bits("2*2", NULL);
	bad_bits("b + 1", NULL);
	bad_bits("b + 2*", NULL);

	good_regs("i", 1);
	good_regs(" s", 1);
	good_regs("l ", 2);
	good_regs(" S ", 2);
	good_regs("  f  ", 2);
	good_regs("int64", 4);
	good_regs("I", 4);
	good_regs("d", 4);
	good_regs("skip", 1);	// this is questionable
	good_regs(" l*3 + 5 * d + skip + skip + skip * 10 +float", 2*3+5*4+1+1+1*10+2);

	bad_regs("blablabla", NULL);
}
```
Compile and run:
```
$ cc test.c src/datatype.c -Ipath/to/valid/zabbix/include && ./a.out
```

File: src/modbus.c [hunk 1]
```diff
@@ -281,74 +234,58 @@ int zbx_modbus_read_registers(AGENT_REQUEST *request, AGENT_RESULT *result)
...
-			datatype = MODBUS_SIGNED_INT;
-		else if (!strcmp(MODBUS_UINT32_STR, param5))
-			datatype = MODBUS_UINT32;
-		else if (!strcmp(MODBUS_SIGNED_INT32_STR, param5))
-			datatype = MODBUS_SIGNED_INT32;
-		else if (!strcmp(MODBUS_FLOAT_STR, param5))
-			datatype = MODBUS_FLOAT;
-		else if (!strcmp(MODBUS_SIGNED_INT64_STR, param5))
-		{
-			SET_MSG_RESULT(result, strdup("datatype 'int64' is not supported."));
```
Comment (v-zhuravlev) [inline src/modbus.c]: Removing this breaks ModbusErrorsTestCase.test_int64_datatype 
```
self = <test_modbus_errors.ModbusErrorsTestCase testMethod=test_int64_datatype>
    def test_int64_datatype(self):
        key = "modbus_read_registers["+self.host+",3,14,3,int64]"
>       self.assertEqual(zabbix_get(key),'ZBX_NOTSUPPORTED: datatype \'int64\' is not supported.')
E       AssertionError: '-4629812106207628288.000000' != "ZBX_NOTSUPPORTED: datatype 'int64' is not supported."
tests/test_modbus_errors.py:40: AssertionError
```

Comment (i-ky) [reply src/modbus.c]: We were discussing returning a string in this case. This way we don't loose any information on the way from module to Zabbix. There will be inevitable loss on the way from Zabbix to DB, but at least user can decide what's more important - sign or precision.

File: tests/test_modbus_datatype_errors.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        self.assertEqual(zabbix_get(key), 'ZBX_NOTSUPPORTED: Invalid type in the datatype expression.')
+
+    @unittest.skip("This test currently breaks the module")
+    def test_modbus_datatype_whitespace_0(self):
+        formula = "'     '"
+        key = "modbus_read_registers["+self.host+",1,2,3,"+formula+"]"
+        self.assertEqual(zabbix_get(key), 'ZBX_NOTSUPPORTED: Invalid type in the datatype expression.')
+    
+    @unittest.skip("This test currently breaks the module")
+    def test_modbus_no_datatype_for_read_registers(self):
```
Comment (v-zhuravlev) [inline tests/test_modbus_datatype_errors.py]: In case of empty datatype param - defaults should be used:
> otherwise, defaults will be used:
MODBUS_BIT if modbus function 1 or 2.
MODBUS_UINT16 if modbus_function 3 or 4.
But currently providing empty datatype just makes module crash.

File: tests/test_modbus_datatype_errors.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        formula = "2*skip"
+        key = "modbus_read_registers["+self.host+",1,2,3,"+formula+"]"
+        self.assertEqual(zabbix_get(key), 'ZBX_NOTSUPPORTED: Invalid type in the datatype expression.')
+
+    def test_modbus_datatype_skip_2(self):
+        formula = "'  skip   '"
+        key = "modbus_read_registers["+self.host+",1,2,3,"+formula+"]"
+        self.assertEqual(zabbix_get(key), 'ZBX_NOTSUPPORTED: Invalid type in the datatype expression.')
+
+    @unittest.skip("This test currently breaks the module")
```
Comment (v-zhuravlev) [inline tests/test_modbus_datatype_errors.py]: Not sure whether this should be considered an error or default should be used?

File: src/modbus.c (hunk 1 above)
Comment (v-zhuravlev) [reply src/modbus.c]: Ok, I changed this test to accept more generic error message.

**Comment by v-zhuravlev** [comment]
Here is what left at this point (correct me if I'm wrong):

- [ ] parse formula of just 'skip'
- [ ] handle empty or whitespace datatype expression properly
- [ ] clang-format test fix
- [ ] add the special error message when `bit` is used in the formula.

File: tests/test_modbus_datatype_errors.py (hunk 2 above)
Comment (i-ky) [reply tests/test_modbus_datatype_errors.py]: Should be fixed by 5ff5080899fc9d2b462bdf274c69a37013a449c5, please double check.

Comment (v-zhuravlev) [reply tests/test_modbus_datatype_errors.py]: also these units test for empty datatypes:  `modbus_no_datatype_for_read_coils`, `modbus_no_datatype_for_read_registers`
